tempo_embeddings/text/abstractcorpus.py

```diff
@@ -0,0 +1,296 @@
+from functools import lru_cache
+from typing import Any
+from typing import Iterable
+from typing import Optional
+import numpy as np
+from numpy.typing import ArrayLike
+from scipy.sparse import csr_matrix
+from sklearn.cluster import HDBSCAN
+from sklearn.feature_extraction.text import TfidfVectorizer
+from .highlighting import Highlighting
+from .passage import Passage
+
+
+class AbstractCorpus(ABC):
+    @property
+    @abstractmethod
+    def passages(self) -> list[Passage]:
+        return NotImplemented
+
+    @property
+    @abstractmethod
+    def embeddings(self) -> Optional[ArrayLike]:
+        return NotImplemented
+
+    @property
+    def label(self) -> Optional[str]:
+        return self._label
+
+    @label.setter
+    def label(self, value: str):
+        self._label = value
+
+    def __eq__(self, other: object) -> bool:
+        return other._passages == self._passages
+
+    def __len__(self) -> int:
+        return len(self._passages)
+
+    def __contains__(self, passage: Passage) -> bool:
+        return passage in self._passages
+
+    def _validate_embeddings(self) -> None:
+        if self._embeddings is not None and self._embeddings.shape[0] != len(
+            self._passages
+        ):
+            raise ValueError(
+                f"Embeddings shape is {self._embeddings.shape}, "
+                f"but number of passages is {len(self._passages)}."
+            )
+
+    def centroid(self) -> ArrayLike:
+        """The mean for all passage embeddings."""
+        return np.array(self.embeddings).mean(axis=0)
+
+    def texts(self) -> Iterable[str]:
+        return (passage.text for passage in self._passages)
+
+    @property
+    def highlightings(self) -> list[Highlighting]:
+        return [passage.highlighting for passage in self.passages]
+
+    def metadata_fields(self) -> set[str]:
+        """Returns all metadata fields in the corpus."""
+
+        return {key for passage in self.passages for key in passage.metadata}
+
+    def has_metadata(self, key: str, strict=False) -> bool:
+        """Returns True if the corpus has a metadata key.
+
+        Args:
+            key: The metadata key to check for.
+            strict: If True, returns True only if all passages have the key.
+        """
+        condition = all if strict else any
+        return condition(key in passage.metadata for passage in self.passages)
+
+    def get_metadatas(self, key: str) -> list[Any]:
+        """Returns all metadata values for a key."""
+
+        return [passage.metadata.get(key) for passage in self.passages]
+
+    def set_metadatas(self, key, value):
+        """Sets a metadata key to a value for all passages.
+
+        Args:
+            key: The metadata key to set.
+            value: The value to set the metadata key to.
+        """
+
+        for passage in self.passages:
+            passage.set_metadata(key, value)
+
+    def hover_datas(self, metadata_fields=None) -> list[dict[str, str]]:
```

Comment:
Indeed! Although I prefer the `list` type hint because `typing.List` has been deprecated since Python 3.9: https://docs.python.org/3/library/typing.html#typing.List

Reply:
I knew that `list` type doesn't support subscription until py39. But since you decide to raise the lower bound of Python version up to 3.9, this indeed won't be an issue.